File: 0396-rotate-function/0396-rotate-function.py
class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        value = sum(i*n for i,n in enumerate(nums))
        total = sum(nums)
        best = value
        
        for n in nums[:0:-1]:
            value += total
            value -= len(nums)*n
            best = max(best, value)
        
        return best

File: 0396-rotate-function/0396-rotate-function.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxRotateFunction = function(nums) {
    let sum = 0;
    let val = 0;
    for (let i=0;i<nums.length;i++) {
        val += i * nums[i];
        sum += nums[i];
    }
    let best = val;
    
    for (let j=nums.length-1;j>0;j--) {
        val += sum;
        val -= nums.length * nums[j];
        best = Math.max(best, val)
    }
    
    return best
    
};